docs: backend deployment documentation (#192)
Substra/substra-documentation: docs/source/operations/howto/backend-deployment.rst

*****************************
How to deploy Substra Backend
*****************************

This guide shows you how to deploy the backend component of Substra.

Prerequisites
=============

To deploy a Substra backend you will need a fully configured Kubernetes cluster.
You will also need to install `Helm <https://helm.sh/>`_.

Preparing your Helm values
==========================

The Substra backend deployment is packaged using Helm.
You can find the complete description of values that can be used to configure the chart on `Artifact Hub <https://artifacthub.io/packages/helm/substra/substra-backend>`_.

To configure your values:

#. Add the Helm repository for Substra:

   .. code-block:: bash

      helm repo add substra https://substra.github.io/charts/

#. Create a Helm values file named ``backend-values.yaml`` with the following content:

   .. code-block:: yaml

>       organizationName: ORGANIZATION

Reviewer: For deploying multiple backend in a channel, should we just repeat this how-to guide with different `organizationName` or are there other steps?

Author: It's mostly that but some steps are still undocumented at the moment like how to connect two backend together for model exchange.

Reviewer: I think it would be worthwhile to add those steps (provided it remains a reasonable effort). In most use cases multiple backends are deployed, so this additional part would definitely be very useful.

Author: Yes, it's just that I prefer to go step by step in order to have reasonable amounts of docs for you to review ! :)

Reviewer: Makes sense, thank you!

#. Configure your Substra backend Ingress. In the ``backend-values.yaml`` file add the following content:

   .. code-block:: yaml

      config:
        ALLOWED_HOSTS: '[".HOSTDOMAIN"]'

      server:
        defaultDomain: https://SUBDOMAIN.HOSTDOMAIN:443
        commonHostDomain: HOSTDOMAIN

        ingress:
          enabled: true
          hostname: SUBDOMAIN.HOSTDOMAIN
   
   | Replace ``HOSTDOMAIN`` with the domain of your server without the lowest subdomain (e.g. for a server exposed at ``api.substra.org`` ``HOSTDOMAIN`` would be ``substra.org``).
   | Replace ``SUBDOMAIN`` with the lowest subdomain (e.g. for a server exposed at ``api.substra.org`` ``SUBDOMAIN`` would be ``api``).

#. Configure your connection to the orchestrator. In the ``backend-values.yaml`` file add the following content:

   .. code-block:: yaml

      orchestrator:
        host: ORCHESTRATOR_HOSTNAME
        port: ORCHESTRATOR_PORT
        mspID: ORGANIZATION

   | Replace ``ORCHESTRATOR_HOSTNAME`` with the hostname of the orchestrator.
   | Replace ``ORCHESTRATOR_PORT`` with the port of your orchestrator (Should be ``80`` if TLS is disabled, otherwise ``443``).

>    | Replace ``ORGANIZATION`` with the name of your Organization. It should be the same value as for the ``organizationName`` key.

Reviewer 2: Note for later: we should eventually remove this `orchestrator.mspID` value if it's the same than the top level `organization`.

Author: It can be different but for simplicity it's almost always the same.

.. _backend-channel-config:

#. Configure your :term:`Substra Channels <Channel>`. 
   In the ``backend-values.yaml`` add the following content under the ``orchestrator`` key:

   .. code-block:: yaml

      channels:
        - CHANNEL:
            restricted: RESTRICTED
            model_export_enabled: MODEL_EXPORT
            chaincode:
              name: mycc

   | Replace ``CHANNEL`` with the name of a channel you want to be part of, it should match one of the channels defined in your :ref:`Orchestrator Substra Channels <orchestrator-channel-config>`.

>    | Replace ``RESTRICTED`` with ``true`` if you should be the only member of this channel else ``false``.

Reviewer 2: ```suggestion
   | Replace ``RESTRICTED`` with ``true`` if your organization should be the only member of this channel else ``false``.
```

>    | Replace ``MODEL_EXPORT`` with ``true`` if you want users from this channel to be able to download models produced by the platform, else ``false``.

Reviewer 2: Nit, I'd rather have a passive form:
```suggestion
   | Replace ``MODEL_EXPORT`` with ``true`` to allow users from this channel to be able to download models produced by the platform, else ``false``.
```

> #. Optional: If your Orchestrator have TLS enabled:

Reviewer 3: ```suggestion
#. Optional: If your Orchestrator has TLS enabled:
```

   #. Retrieve the CA certificate from your orchestrator:

      The CA certificate is the ``orchestrator-ca.crt`` file generated at the :ref:`Generate your Certificate Authority certificate <orchestrator-cacert-generation>` step of the Orchestrator deployment.
      If a public Certificate Authority was used to generate the orchestrator certificate you will need to fetch the certificate of the Certificate Authority.

   #. Create a ConfigMap containing the CA certificate:

      .. code-block:: bash

         kubectl create configmap orchestrator-cacert --from-file=ca.crt=orchestrator-ca.crt

   #. Configure your backend to enable Orchestrator TLS. In the ``backend-values.yaml`` file add the following content under the ``orchestrator`` key:

      .. code-block:: yaml

           tls:
             enabled: true
             cacert: orchestrator-cacert

> #. Add users to your backend. In the ``backend-values.yaml`` file add the following content:

Reviewer 2: Do we have a User concept definition we can link? It might not be obvious that users are org-scoped and that it conditions their API usage.

Author: Good point, I'll see what I can find. Especially since now there is a user management "tool" in the frontend.

   .. code-block:: yaml

      addAccountOperator:
        users:
          - name: USERNAME
            secret: PASSWORD
            channel: CHANNEL

   | Replace ``USERNAME`` with the name of the user you want to add.
   | Replace ``PASSWORD`` with the password of the user you want to add. It should be at least 20 characters long.
   | Replace ``CHANNEL`` with the name of the channel this user is part of. It should match one of the channels defined in your :ref:`Substra Channel configuration <backend-channel-config>`.

Deploy the Chart
================

> To deploy the Substra Backend chart in your Kubernete cluster follow these steps:

Reviewer 4: Small typo: kubernete**s** (s missing)

> #. Deploy the orchestrator Helm chart:

Reviewer 5: there's a typo, it's the backend helm chart
